org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractFoldingTest.java

```diff
@@ -0,0 +1,134 @@
+/**
+ * Copyright (c) 2019, 2020 itemis AG (http://www.itemis.eu) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.xtext.ui.testing;
+
+import java.lang.reflect.InvocationTargetException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.regex.Pattern;
+
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.jface.text.BadLocationException;
+import org.eclipse.xtext.resource.FileExtensionProvider;
+import org.eclipse.xtext.ui.XtextProjectHelper;
+import org.eclipse.xtext.ui.editor.folding.FoldedPosition;
+import org.eclipse.xtext.ui.editor.folding.IFoldingRegionProvider;
+import org.eclipse.xtext.ui.editor.model.IXtextDocument;
+import org.eclipse.xtext.ui.testing.util.IResourcesSetupUtil;
+
+import com.google.common.annotations.Beta;
+import com.google.common.base.Joiner;
+import com.google.inject.Inject;
+
+/**
+ * @author miklossy - Initial contribution and API
+ * @since 2.19
+ */
+@Beta
+public abstract class AbstractFoldingTest extends AbstractEditorTest {
+
+	@Inject
+	protected FileExtensionProvider fileExtensionProvider;
+
+	@Inject
+	protected IFoldingRegionProvider foldingRegionProvider;
+
+	/**
+	 * @return the special symbol indicating the beginning of a folding region.
+	 */
+	public String b() {
+		return "[>";
+	}
+
+	/**
+	 * @return the special symbol indicating the end of a folding region.
+	 */
+	public String e() {
+		return "<]";
+	}
+
+	/**
+	 * Test that the expected folding regions are present on a given DSL text.
+	 * 
+	 * @param charSequence
+	 *            The DSL text where the expected folding regions are enclosed by the {@link #b() b} and {@link #e() e} symbols.
+	 */
+	public void testFoldingRegions(CharSequence charSequence) {
+		foldingRegionsArePresent(openInEditor(dslFile(charSequence)), charSequence);
+	}
+
+	protected IFile dslFile(CharSequence text) {
+		try {
+			IFile file = IResourcesSetupUtil.createFile(getProjectName(), getFileName(), getFileExtension(), getContent(text));
+
+			/*
+			 * TODO: find a better (with good performance) solution
```

Comment:
The **TODO** was already there. Please suggest if there is a better way to deal with this situation. I'll make the required change.